cpp/src/transpose/transpose.cu

```diff
@@ -0,0 +1,176 @@
+ * 
+ * @tparam T  Datatype of values pointed to by the views
+ * @param input[in]  Device view of input columns' data
+ * @param output[out]  Mutable device view of pre-allocated output columns' data
+ */
+template <typename T>
+__global__
+void gpu_transpose(table_device_view const input, mutable_table_device_view output)
+{
+  size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  size_type stride_x = blockDim.x * gridDim.x;
+  size_type stride_y = blockDim.y * gridDim.y;
+
+  for (size_type i = x; i < input.num_columns(); i += stride_x) {
+    for (size_type j = y; j < input.num_rows(); j += stride_y) {
+      output.column(j).element<T>(i) = input.column(i).element<T>(j);
+    }
+  }
+}
+
+/**
+ * @brief Transposes the null mask from ncols x nrows input to nrows x ncols output
+ * 
+ * @tparam T  Datatype of values pointed to by the views
+ * @param input[in]  Device view of input columns' data
+ * @param output[out]  Mutable device view of pre-allocated output columns' data
+ */
+__global__
+void gpu_transpose_valids(table_device_view const input, mutable_table_device_view output)
+{
+  size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  size_type stride_x = blockDim.x * gridDim.x;
+  size_type stride_y = blockDim.y * gridDim.y;
+
+  size_type i = x;
+  auto active_threads = __ballot_sync(0xffffffff, i < input.num_columns());
+  while (i < input.num_columns()) {
+    for (size_type j = y; j < input.num_rows(); j += stride_y) {
+      auto const result = __ballot_sync(active_threads, input.column(i).is_valid(j));
+
+      // Only one thread writes output
+      if (0 == threadIdx.x % WARP_SIZE) {
+        output.column(j).set_mask_word(word_index(i), result);
+      }
+    }
+
+    // Update active threads before branching
+    i += stride_x;
+    active_threads = __ballot_sync(active_threads, i < input.num_columns());
+  }
+}
+
+struct launch_kernel {
+  template <typename T>
+  void operator()(
+    table_view const& input,
+    mutable_table_view& output,
+    bool has_null,
+    cudaStream_t stream)
+  {
+    // Copy input columns `data` and `valid` pointers to device
+    auto device_input = table_device_view::create(input, stream);
+    auto device_output = mutable_table_device_view::create(output, stream);
+
+    dim3 dimBlock(WARP_SIZE, WARP_SIZE);
+    dim3 dimGrid(std::min(util::div_rounding_up_safe(input.num_columns(), WARP_SIZE), MAX_GRID_SIZE),
+                 std::min(util::div_rounding_up_safe(input.num_rows(), WARP_SIZE), MAX_GRID_SIZE));
+
+    gpu_transpose<T><<<dimGrid, dimBlock, 0, stream>>>(*device_input, *device_output);
+
+    if (has_null) {
+      gpu_transpose_valids<<<dimGrid, dimBlock, 0, stream>>>(*device_input, *device_output);
+
+      // Force null counts to be recomputed next time they are queried
+      for (auto& column : output) {
+        column.set_null_count(UNKNOWN_NULL_COUNT);
```

Comment:
Unfortunately this isn't doing what you expect.

It currently isn't possible to modify the null count of a `cudf::column`. via a `mutable_column_view`. This would require the `view` to have a pointer/reference to the column's null count in order to be able to modify it.

Luckily, anytime you create a `mutable_column_view` from a `column`, it automatically invalidates the null count. So this step isn't necessary.

Reply:
Furthermore, I think the `gpu_transpose_valids` kernel should be updated to also compute the null count for each output column. Otherwise, if there is 100k output columns, it would require 100k kernels to get the null count of each one.

Reply:
Oh, I see. The original column and the view each keep a separate tally of the nulls and the view has no way of setting the null count of the original column. Yikes, that was not obvious to me.

Ok, I think I'll need to modify launch_kernel so that it does the allocation of the output columns so that I have the actual columns rather than mutable views in that scope.

Reply:
I added the null count computation and updated the unit tests to check it. There's a consequence though: it requires synchronization after copying the null counts from the device back to the host.

Reply:
> Yikes, that was not obvious to me.

Yeah, it's definitely a wart that I hope to eventually be able to address.

Reply:
Even though it is a wart, the good news it won't lead to incorrect behavior, just worse performance. This is because creating a `mutable_view` to a column will always invalidate it's internal null count, which will force revaluation on the next invocation of `null_count()`.